File: leetcode/indexed tree/304. Range Sum Query 2D - Immutable/py/main.py
class NumMatrix:
    # Time O(N*M)
    # Space O(1)
    def __init__(self, matrix):
        if len(matrix):
            n, m = len(matrix), len(matrix[0])

            # calc prefix sum for each row
            for i in range(n):
                for j in range(1, m):
                    matrix[i][j] += matrix[i][j - 1] 

            # calc prefix sum for each col
            for j in range(m):
                for i in range(1, n):
                    matrix[i][j] += matrix[i - 1][j]

            self.matrix = matrix

    """
    Let's calculate sum for region row1=1, col1=1, row2=2, col2=2

    Take sum of the region matrix[2][2] -> 21
    
    Substract the sum of region matrix[0][2] -> 4
    Substract the sum of region matrix[2][0] -> 9

    But we've substracted twice the sum of matrix[0][0]

    So we have to add matrix[0][0] to answer

    3, 3,   4
    8, 14,  18
    9, 17 , 21

    """
    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        current = self.matrix[row2][col2]
        origin = self.matrix[row1 - 1][col1 - 1] if col1 > 0 and row1 > 0 else 0
        left = self.matrix[row2][col1 - 1] if col1 > 0 else 0
        right = self.matrix[row1 - 1][col2] if row1 > 0 else 0
        
        return current - left - right + origin

File: leetcode/indexed tree/304. Range Sum Query 2D - Immutable/js/main.js
class NumMatrix {
  constructor(matrix) {
    if (matrix.length) {
      let n = matrix.length;
      let m = matrix[0].length;

      this.prefix_sum = Array(n + 1)
        .fill(0)
        .map(() => Array(m + 1).fill(0));

      for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
          this.insert(i, j, matrix[i][j]);
        }
      }
    }
  }

  getNext(i) {
    return i + (i & -i);
  }

  getPrev(i) {
    return i - (i & -i);
  }

  insert(x, y, val) {
    for (let i = x + 1; i < this.prefix_sum.length; i = this.getNext(i)) {
      for (let j = y + 1; j < this.prefix_sum[0].length; j = this.getNext(j)) {
        this.prefix_sum[i][j] += val;
      }
    }
  }

  search(x, y) {
    let sum = 0;

    for (let i = x; i > 0; i = this.getPrev(i)) {
      for (let j = y; j > 0; j = this.getPrev(j)) {
        sum += this.prefix_sum[i][j];
      }
    }

    return sum;
  }

  sumRegion(row1, col1, row2, col2) {
    return this.search(row2 + 1, col2 + 1) - this.search(row1, col2 + 1) - this.search(row2 + 1, col1) + this.search(row1, col1);
  }
}
